Added log environment variables

qdm12 opened the pull request:
- Added ~`LOGVERBOSE`~ `LOG_VERBOSE` and ~`LOGDEBUG`~ `LOG_DEBUG` environment variables
- ~`/output/usr/local/bin/*` files are only readable and executable by their owner~
- ~`/output/usr/local/bin/*` files are owned by $SUID at launch (entrypoint) as $SUID is user-defined~

qdm12 committed:
Added verbose env variables

frebib commented:
I'll have a look at this tomorrow, it's been hectic today

frebib commented on README.md:
```diff
@@ -59,6 +59,8 @@ Finally, navigate to [app.plex.tv/desktop](https://app.plex.tv/desktop) or [your
 - `$ADVERTISE_IP`         - This variable defines the additional IPs on which the server may be be found. For example: `http://10.1.1.23:32400`. This adds to the list where the server advertises that it can be found.
 - `$DISABLE_REMOTE_SEC`   - 
 - `$PLEX_CLAIM`           - The claim token for the server to obtain a real server token. If not provided, server will not be automatically logged in. If server is already logged in, this parameter is ignored.
+- `$LOGDEBUG`             - Disables debug logging if set to 0, and defaults to 1 (enables debug logging).
```
Pedantic but can these variables have an underscore in, e.g. `LOG_DEBUG`

qdm12 replied:
The [Plex options](https://support.plex.tv/articles/201105343-advanced-hidden-server-settings/) are `logDebug` and `LogVerbose` without spaces, so should we add an underscore?

frebib replied:
Typically a change in case denotes a different word. The equivalent would be an underscore in UPPER_SNAKE_CASE

frebib commented on entrypoint:
```diff
@@ -51,6 +51,26 @@ if [ ! -e "\$PREF_FILE" ]; then
 fi
 
+# Changes the log verbosity
+if [ "$LOGVERBOSE" = "1" ] || [ -z "$LOGVERBOSE" ]; then
```
I've been thinking about this and I'm not sure how to handle "truthy" values. Whether that is any non-empty value or some specific truthy values like `true`, `1` etc. Either way as is I don't think it'll work.

Perhaps something like this

```
bool() {
    case $(echo "$1" | tr 'A-Z' 'a-z') in
        1|t|true) return 0;;
        *) return 1;;
    esac;
}

if bool "$LOG_VERBOSE"; then
    ...
```

qdm12 replied:
I am just using the Plex built-in options, which are `0` or `1` for `LogVerbose`. I use `[ -z "$LOGVERBOSE" ]` as a default value (keep it as it is: `1`)

frebib commented on entrypoint:
```diff
@@ -51,6 +51,26 @@ if [ ! -e "\$PREF_FILE" ]; then
+if [ "$LOGVERBOSE" = "1" ] || [ -z "$LOGVERBOSE" ]; then
+    setPref "LogVerbose" "1"
+elif [ "$LOGVERBOSE" = "0" ]; then
+    setPref "LogVerbose" "0"
```
As for setting the inverse value, perhaps we should only do that if `LOG_VERBOSE` is set/non-empty.

Following the above, something like this would work I think:
```
if [ -n "$LOG_VERBOSE" ]; then
    if bool "$LOG_VERBOSE"; then
        setPref "LogVerbose" "1"
￼    else
￼         setPref "LogVerbose" "0"
    fi
fi
```
or of course mapping the true/false return value of the bool function to a 0/1 and setting that as the value would probably be better :)

qdm12 replied:
Again, I think it's best to stick to Plex built-in values `0` or `1`, and defaults to `1` (as it is now).

frebib commented on entrypoint:
```diff
@@ -51,6 +51,26 @@ if [ ! -e "\$PREF_FILE" ]; then
+fi
+
+# Changes the debug verbosity
+if [ "$LOGDEBUG" = "1" ] || [ -z "$LOGDEBUG" ]; then
```
Same as above, otherwise good

frebib commented:
Can I ask why you changed the permission of the executable files? Was it to ensure the config file gets written as the correct user? If so we should handle that by attempting to chown the file after creation. It's useful for the scripts to be executable by anyone, and not really a risk to do so, for the case of debugging.

qdm12 commented:
> Can I ask why you changed the permission of the executable files?

The executable files had `755` permissions and were owned by root I believe:
- Writable, readable and executable as root (not needed as root can change them anyway)
- Readable and executable as group root (not needed)
- Readable and executable as all other users (that's what was used with SUID)

I think it should be restricted to SUID (`chown $SUID`) only and be only readable and executable (`chmod 500`).

> It's useful for the scripts to be executable by anyone, and not really a risk to do so, for the case of debugging.

It the user wants to modify executables, he can `docker exec -it plex /bin/sh`

> Was it to ensure the config file gets written as the correct user? If so we should handle that by attempting to chown the file after creation.

At each docker run / start, the user provided SUID may change so `chown` has to be in the entrypoint, not in the Dockerfile.

frebib commented:
> The executable files had 755 permissions and were owned by root

This is pretty typical of an executable in `$PREFIX/bin`. Very rarely/never do you use anything less than 555 those situations.
Regardless, I feel like changing the permissions of the script is out of the scope of this PR. You should drop those commits

qdm12 commented:
It is out the scope indeed. ~But if the container gets comprised and an attacker runs as user SUID, it can change the executables and run them which is unwanted I think, please correct me if I'm wrong.~ I'm wrong, as it's owned by root. So never mind, keep it like that sorry for messing around! ~Maybe you could add it in your next commit sometime?~ I'll remove it from the PR soon. Thanks!

frebib commented:
Thanks for reverting the SUID stuff.

I need to have a think about default values because one of the things we have to be careful of is setting a default value in the env. If a user changes it from the Plex UI and restarts the container it will always be overwritten which seems counter intuitive and can add a source of confusion. IMO there should be no default value and it should handle the empty case too.

One other thing is it should `exit 1` on an error to be clear that startup was unsuccessful

frebib commented:
Apologies for being so pedantic btw

qdm12 commented:
It should be fixed, let me know if I should change anything.

qdm12 commented:
> I need to have a think about default values because one of the things we have to be careful of is setting a default value in the env. If a user changes it from the Plex UI and restarts the container it will always be overwritten which seems counter intuitive and can add a source of confusion. IMO there should be no default value and it should handle the empty case too.

Totally agreed, I'll find a good solution for that.

> One other thing is it should exit 1 on an error to be clear that startup was unsuccessful

I would totally agree a week before, but I stumbled on a program which would exit as 0 because it was not a program error but a successful run of the program (although it stopped). Since then I changed my mind about exit codes ! But I understand your (my previous) opinion, which makes sense too... I'll make it a `1` 😄

qdm12 committed:
PR changes requested

- `LOG_DEBUG` or `LOG_VERBOSE` do not change anything it they are unset (same as before)
- `LOG_DEBUG` or `LOG_VERBOSE` take precedence on Plex configuration file at start when set to `0` or `1`. The user can change the Preferences once the container is started which will overwrite the environment variables.

frebib commented:
lgtm

@Adam-Ant?

Adam-Ant commented on README.md:
```diff
@@ -59,6 +59,8 @@ Finally, navigate to [app.plex.tv/desktop](https://app.plex.tv/desktop) or [your
 - `$ADVERTISE_IP`         - This variable defines the additional IPs on which the server may be be found. For example: `http://10.1.1.23:32400`. This adds to the list where the server advertises that it can be found.
 - `$DISABLE_REMOTE_SEC`   - 
 - `$PLEX_CLAIM`           - The claim token for the server to obtain a real server token. If not provided, server will not be automatically logged in. If server is already logged in, this parameter is ignored.
+- `$LOGDEBUG`             - Disables debug logging if set to 0, and defaults to 1 (enables debug logging).
```
These need to be updated to have the underscore in them like the code now does. Could there also be a note somewhere in the readme saying that these will override the Plex UI options for debug and verbose.

frebib replied:
@qdm12

qdm12 replied:
Hi there, sorry for forgetting to update the doc. I'll do it today/tomorrow. Thanks for letting me know!

Adam-Ant reviewed (changes requested):
Code and implementation LGTM, just a few tweaks needed to the documentation.

qdm12 committed:
Fixed readme.md with additional information on LOG_DEBUG and LOG_VERBOSE

Adam-Ant commented:
Thanks for those changes, now LGTM. Merging!
